### PR TITLE
ceph: add an rbd-mirror bootstrap token on cluster creation 

### PR DESCRIPTION
**Pull request by leseb**
Signed-off-by: Sébastien Han <seb@redhat.com>

**Description of your changes:**

**Which issue is resolved by this Pull Request:**
Resolves #

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/master/development-flow.html#commit-structure).
- [ ] **Skip Tests for Docs**: Add the flag for skipping the build if this is only a documentation change. See [here](https://github.com/rook/rook/blob/master/INSTALL.md#skip-ci) for the flag.
- [ ] **Skip Unrelated Tests**: Add a flag to run tests for a specific storage provider. See [test options](https://github.com/rook/rook/blob/master/INSTALL.md#test-storage-provider).
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/master/development-flow.html#submitting-a-pull-request)
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.
- [ ] Pending release notes updated with breaking and/or notable changes, if necessary.
- [ ] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [ ] Code generation (`make codegen`) has been run to update object specifications, if necessary.

### CODE REVIEW

**Comment by leseb** [comment]
This depends on https://github.com/rook/rook/pull/8380 since the Secret does not contain the pool name and the "mirror pool peer bootstrap import" requires it. Once merged, the pool will be part of CephBlockPool.

File: pkg/daemon/ceph/client/mirror.go [hunk 1]
```diff
@@ -17,21 +17,41 @@ limitations under the License.
...
+	rbdMirrorPeerCaps      = []string{"mon", "profile rbd-mirror-peer", "osd", "profile rbd"}
+	rbdMirrorPeerKeyringID = "rbd-mirror-peer"
 )
 
 // ImportRBDMirrorBootstrapPeer add a mirror peer in the rbd-mirror configuration
 func ImportRBDMirrorBootstrapPeer(context *clusterd.Context, clusterInfo *ClusterInfo, poolName, direction string, token []byte) error {
 	logger.Infof("add rbd-mirror bootstrap peer token for pool %q", poolName)
 
 	// Token file
+	// TODO: use mktemp?
```
Comment (travisn) [inline pkg/daemon/ceph/client/mirror.go:54]: planning on this TODO?

**Review by travisn — COMMENTED** [review]
Do we need to update any documentation for this?

File: pkg/daemon/ceph/client/mirror.go (hunk 1 above)
Comment (leseb) [reply pkg/daemon/ceph/client/mirror.go:54]: Not in this PR. https://github.com/rook/rook/issues/8446

**Review by travisn — APPROVED** [review]

File: pkg/daemon/ceph/client/mirror.go [hunk 2]
```diff
@@ -17,21 +17,41 @@ limitations under the License.
...
+	PoolID    int    `json:"pool_id"`
+	Namespace string `json:"namespace"`
+}
+
+var (
+	rbdMirrorPeerCaps      = []string{"mon", "profile rbd-mirror-peer", "osd", "profile rbd"}
+	rbdMirrorPeerKeyringID = "rbd-mirror-peer"
 )
 
 // ImportRBDMirrorBootstrapPeer add a mirror peer in the rbd-mirror configuration
```
Comment (sp98) [inline pkg/daemon/ceph/client/mirror.go:49]:
```suggestion
// ImportRBDMirrorBootstrapPeer adds a mirror peer in the rbd-mirror configuration
```

File: pkg/daemon/ceph/client/mirror.go [hunk 3]
```diff
@@ -307,3 +327,57 @@ func ListSnapshotSchedulesRecursively(context *clusterd.Context, clusterInfo *Cl
...
+	4) Enable destination pool for mirroring
+	rbd --cluster site-b mirror pool enable image-pool image
+
+	5) Add the peer details to the destination pool
+
+	6) Repeat the steps flipping source and destination to enable
+	bi-directional mirroring
+*/
+func CreateRBDMirrorBootstrapPeerWithoutPool(context *clusterd.Context, clusterInfo *ClusterInfo) ([]byte, error) {
+	logger.Infof("create rbd-mirror bootstrap peer token %q", getQualifiedUser(rbdMirrorPeerKeyringID))
```
Comment (sp98) [inline pkg/daemon/ceph/client/mirror.go:355]: `getQualifiedUser` is called thrice in this method.  Should we call `getQualifiedUser` once at the beginning of the method. Though it just prepends a string.

**Review by sp98 — CHANGES_REQUESTED** [review]
nit

**Review by sp98 — APPROVED** [review]

File: pkg/daemon/ceph/client/mirror.go [hunk 4]
```diff
@@ -17,21 +17,41 @@ limitations under the License.
...
+)
+
+// PeerToken is the content of the peer token
+type PeerToken struct {
+	ClusterFSID string `json:"fsid"`
+	ClientID    string `json:"client_id"`
+	Key         string `json:"key"`
+	MonHost     string `json:"mon_host"`
+	// These fields are added by Rook and NOT part of the output of client.CreateRBDMirrorBootstrapPeer()
+	PoolID    int    `json:"pool_id"`
```
Comment (Madhu-1) [inline pkg/daemon/ceph/client/mirror.go:40]: @leseb can we remove pool_id? this is not much helpful for 3 reason
* As we are going to have mapping of pool Id's
* As this token is per ceph cluster we will have multiple pool_id's in a cluster?
* And when this token is generated the pool ID will always be empty?

File: pkg/daemon/ceph/client/mirror.go [hunk 5]
```diff
@@ -307,3 +327,58 @@ func ListSnapshotSchedulesRecursively(context *clusterd.Context, clusterInfo *Cl
...
+	logger.Infof("create rbd-mirror bootstrap peer token %q", fullClientName)
+	key, err := AuthGetOrCreateKey(context, clusterInfo, fullClientName, rbdMirrorPeerCaps)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to create rbd-mirror peer key %q", fullClientName)
+	}
+	logger.Infof("successfully created rbd-mirror bootstrap peer token for cluster %q", clusterInfo.NamespacedName().Name)
+
+	mons := util.NewSet()
+	for _, mon := range clusterInfo.Monitors {
+		mons.Add(mon.Endpoint)
```
Comment (Madhu-1) [inline pkg/daemon/ceph/client/mirror.go:365]: @leseb if there is a new mon added or mon IP changes will this token will also get updated?

**Review by Madhu-1 — COMMENTED** [review]
@leseb are we creating a new secret that contains the token per cluster? when ceph cluster is created? i don't see any call  to create secret (i might be missing it sorry for late comment on this one)

File: pkg/daemon/ceph/client/mirror.go (hunk 5 above)
Comment (leseb) [reply pkg/daemon/ceph/client/mirror.go:365]: Yes!

File: pkg/daemon/ceph/client/mirror.go (hunk 4 above)
Comment (leseb) [reply pkg/daemon/ceph/client/mirror.go:40]: I'll remove that in a follow-up PR.

**Pull request merged**